src/libmongoc/tests/test-mongoc-client-pool.c

```diff
@@ -405,6 +405,70 @@ test_client_pool_create_unused_session (void *context)
    ASSERT_NO_CAPTURED_LOGS ("mongoc_client_pool_destroy");
 }
 
+
+/* Tests case where thread is blocked waiting for a client to be pushed back
+ * into the client pool.  Specifically this tests that the program terminates.
+ * Addresses CDRIVER-3757 */
+#ifdef BSON_OS_UNIX
+typedef struct pool_timeout {
+   mongoc_client_pool_t *pool;
+   bson_mutex_t mutex;
+   mongoc_cond_t cond;
+   int nleft;
+} pool_timeout_args_t;
+
+static void *
+worker (void *arg)
+{
+   pool_timeout_args_t *args = arg;
+   mongoc_client_t *client = mongoc_client_pool_pop (args->pool);
+   BSON_ASSERT (client);
+   sleep (1);
```

Comment:
IIUC this sleep is here to make the test fail more consistently, to make it likely that both threads attempt to pop a client at the same time.

I'd argue for removing one of the worker threads and removing this sleep to avoid a slow test:
1. main thread creates client pool with maxPoolSize=1
2. main thread pops a client from pool
3. main thread creates worker thread, which attempts to pop a client but must wait
4. main thread pushes client onto pool

Reply:
Yes, that's the reason!  It is to make it likely that thread2 attempts to the pop a client before the thread1 pushes its client back into the pool, causing thread2 to block forever in its call to `mongoc_client_pool_pop`.  I totally agree that this is a slow test.

I attempted to do what you described and encountered two issues:

1. The main thread seems to push its client before the worker thread attempts to pop a client (which is not the case that I would like to test), unless the main thread sleeps before pushing its client.
2. In the event that the worker thread attempts to pop a client before the main thread pushes its client, the main thread will block forever in its call to `mongoc_client_pool_push`, which is another manifestation of the bug.  The issue with that is that it then requires a different timeout mechanism.

I can do those things but I'm not sure they would necessarily fix the issue of the slow test, because it would still require a call to sleep, and it would also complicate the timeout mechanism (only a little bit).

To make the test faster I have replaced `sleep` with `_umongoc_sleep`, which allows a higher level of control on the sleep time.  Now instead of sleeping for ~2 seconds the test will sleep for ~20 micro seconds in the case where the test passes.  

Really the issue here is that I'm not using the clients for anything, which allows the threads to pop and push before the other threads can pop.  In general I'm not a fan of using `sleep` to try and control the order that things happen in, but in this case I don't mind it that much, because I view it as a simulation of doing something with the client.

In any case, using `_umongoc_sleep` definitely speeds it up.  I still think your idea may be better.  What do you think?

Reply:
> 1. The main thread seems to push its client before the worker thread attempts to pop a client (which is not the case that I would like to test), unless the main thread sleeps before pushing its client.

Good point. You could synchronize the worker thread to signal to the main thread with a condition variable after popping a client.

> 2. In the event that the worker thread attempts to pop a client before the main thread pushes its client, the main thread will block forever in its call to mongoc_client_pool_push, which is another manifestation of the bug. The issue with that is that it then requires a different timeout mechanism.

That's true. If the failure case of this test is an infinite block, that seems much worse than a slightly slow / non-deterministic test.